### PR TITLE
[DPE-4264] [DPE-4948][DPE-4950] Support backups SC + re-enable RS backup test + fix charmcraft issues

### PR DESCRIPTION
**Pull request by MiaAltieri**
## Issue
1. Backup tests are disabled for replica set
2. No evidence that backups are functioning on sharded cluster
3. Newest version of charmcraft has a lot of issues

## Solution
1. Re-enable tests and make necessary changes so that tests may run successful for backups on replica set deployments
2. Add in basic integration tests for backups on sharded cluster
3. Use DP workflow that relies on older charmcraft version

### CODE REVIEW

File: tests/integration/backup_tests/helpers.py [hunk 1]
```diff
@@ -13,29 +12,6 @@
 TIMEOUT = 10 * 60
 
 
-async def destroy_cluster(ops_test: OpsTest, cluster_name: str) -> None:
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/helpers.py:16]: moved this to reflect @Gu1nness change

File: tests/integration/backup_tests/helpers.py [hunk 2]
```diff
@@ -106,11 +82,11 @@ async def count_failed_backups(db_unit: ops.model.Unit) -> int:
     return failed_backups
 
 
-async def set_credentials(ops_test: OpsTest, cloud: str) -> None:
+async def set_credentials(ops_test: OpsTest, github_secrets, cloud: str) -> None:
     """Sets the s3 crednetials for the provided cloud, valid options are AWS or GCP."""
     # set access key and secret keys
-    access_key = os.environ.get(f"{cloud}_ACCESS_KEY", False)
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/helpers.py:112]: updating to new use of secrets

File: tests/integration/backup_tests/helpers.py [hunk 3]
```diff
@@ -132,3 +108,12 @@ async def insert_unwanted_data(ops_test: OpsTest) -> None:
     test_collection.insert_one({"unwanted_data": "bad data 2"})
     test_collection.insert_one({"unwanted_data": "bad data 3"})
     client.close()
+
+
+async def get_backup_list(ops_test: OpsTest, db_app_name=None) -> str:
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/helpers.py:95]: copied from VM

File: tests/integration/backup_tests/test_sharding_backups.py [hunk 4]
```diff
@@ -0,0 +1,437 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
```
Comment (MiaAltieri) [inline tests/integration/backup_tests/test_sharding_backups.py:3]: copied with minor modifications from VM charm (with @Gu1nness recent improvements to tests)

File: tests/integration/helpers.py [hunk 5]
```diff
@@ -670,3 +670,22 @@ def is_relation_joined(ops_test: OpsTest, endpoint_one: str, endpoint_two: str)
         if endpoint_one in endpoints and endpoint_two in endpoints:
             return True
     return False
+
+
+async def destroy_cluster(ops_test: OpsTest, applications: list[str]) -> None:
```
Comment (MiaAltieri) [inline tests/integration/helpers.py:675]: copied from VM charm

File: tests/integration/sharding_tests/writes_helpers.py [hunk 6]
```diff
@@ -0,0 +1,113 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
```
Comment (MiaAltieri) [inline tests/integration/sharding_tests/writes_helpers.py:2]: copied from VM charm with slight mods

File: tests/integration/helpers.py [hunk 7]
```diff
@@ -109,7 +109,7 @@ async def get_address_of_unit(ops_test: OpsTest, unit_id: int, app_name: str = A
 
 
 async def get_password(
-    ops_test: OpsTest, unit_id: int, username="operator", app_name: str = APP_NAME
+    ops_test: OpsTest, unit_id: int = 0, username="operator", app_name: str = APP_NAME
```
Comment (Gu1nness) [inline tests/integration/helpers.py]:
```suggestion
    ops_test: OpsTest, unit_id: int = 0, username : str = "operator", app_name: str = APP_NAME
```

**Review by Gu1nness — DISMISSED** [review]
One very minor suggestion here (not important, just nitpicking)

**Review by Gu1nness — DISMISSED** [review]

**Comment by Gu1nness** [comment]
@MiaAltieri Just saw that:
```
mongodb-k8s-operator/tests/integration/backup_tests/test_backups.py:155: RuntimeWarning: coroutine 'Model.wait_for_idle' was never awaited
```
in the test failing

File: tests/integration/backup_tests/test_backups.py [hunk 8]
```diff
@@ -130,35 +135,29 @@ async def test_blocked_incorrect_creds(ops_test: OpsTest) -> None:
...
 
     # set correct AWS credentials for s3 storage but incorrect configs
-    await helpers.set_credentials(ops_test, cloud="AWS")
+    await helpers.set_credentials(ops_test, github_secrets, cloud="AWS")
 
     # wait for both applications to be idle with the correct statuses
-    await asyncio.gather(
-        ops_test.model.wait_for_idle(apps=[S3_APP_NAME], status="active"),
-        ops_test.model.wait_for_idle(apps=[db_app_name], status="blocked", idle_period=20),
+    ops_test.model.wait_for_idle(apps=[S3_APP_NAME], status="active")
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py]:
```suggestion
    await ops_test.model.wait_for_idle(apps=[S3_APP_NAME], status="active")
```

File: tests/integration/backup_tests/test_backups.py [hunk 9]
```diff
@@ -362,15 +356,18 @@ async def test_restore(ops_test: OpsTest, continuous_writes_to_db) -> None:
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py:356]: A few lines above (line 352) - There is a reference to a non-existing function: `ha_helpers.count_writes(ops_test)`

File: tests/integration/backup_tests/test_backups.py [hunk 10]
```diff
@@ -467,13 +464,9 @@ async def test_restore_new_cluster(ops_test: OpsTest, continuous_writes_to_db, c
             writes_in_new_cluster == writes_in_old_cluster
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py:464]: A few lines above (line 458) - There is a reference to a non-existing function: `ha_helpers.count_writes(ops_test)`

File: tests/integration/backup_tests/test_backups.py [hunk 11]
```diff
@@ -467,13 +464,9 @@ async def test_restore_new_cluster(ops_test: OpsTest, continuous_writes_to_db, c
...
         ), "new cluster writes do not match old cluster writes after restore"
 
-    # TODO there is an issue with on stop and secrets that need to be resolved before
-    # we can cleanup the new cluster, otherwise the test will fail.
-
-    # await helpers.destroy_cluster(ops_test, cluster_name=NEW_CLUSTER)
+    await destroy_cluster(ops_test, cluster_name=NEW_CLUSTER)
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/test_backups.py]:
```suggestion
    await destroy_cluster(ops_test, applications=[NEW_CLUSTER])
```

File: tests/integration/backup_tests/helpers.py [hunk 12]
```diff
@@ -132,3 +108,12 @@ async def insert_unwanted_data(ops_test: OpsTest) -> None:
     test_collection.insert_one({"unwanted_data": "bad data 2"})
```
Comment (Mehdi-Bendriss) [inline tests/integration/backup_tests/helpers.py]: We should delete this helper. It seems to be a broken precursor (reference to missing functions etc..) of the one you added below

File: tests/integration/ha_tests/application_charm/src/charm.py [hunk 13]
```diff
@@ -72,14 +72,21 @@ def app_peer_data(self) -> Dict:
     @property
     def _database_config(self):
         """Returns the database config to use to connect to the MongoDB cluster."""
-        data = list(self.database.fetch_relation_data().values())[0]
+        # In some tests we want to write directly to mongos, but the config-server does not
+        # support integrations to client applications, so the data to connect is set via config.
+        data = list(self.database.fetch_relation_data().values())
+        if len(data) == 0:
```
Comment (Mehdi-Bendriss) [inline tests/integration/ha_tests/application_charm/src/charm.py]: nit - good candidate for the walrus operator:
```suggestion
        if not (data := list(self.database.fetch_relation_data().values())):
```

File: tests/integration/sharding_tests/writes_helpers.py [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+
+class ProcessRunningError(Exception):
+    """Raised when a process is running when it is not expected to be."""
+
+
+async def remove_db_writes(
+    ops_test: OpsTest,
+    db_name: str,
+    coll_name: str = DEFAULT_COLL_NAME,
+) -> bool:
```
Comment (Mehdi-Bendriss) [inline tests/integration/sharding_tests/writes_helpers.py]:
```suggestion
) -> None:
```

**Review by Mehdi-Bendriss — CHANGES_REQUESTED** [review]
Thanks Mia - I have a few comments

**Review by Mehdi-Bendriss — APPROVED** [review]

**Review by Gu1nness — APPROVED** [review]

**Pull request merged**